firestore/bulkwriter.go

```diff
@@ -0,0 +1,319 @@
+package firestore
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	vkit "cloud.google.com/go/firestore/apiv1"
+	"google.golang.org/api/support/bundler"
+	pb "google.golang.org/genproto/googleapis/firestore/v1"
+)
+
+const (
+	// maxBatchSize is the max number of writes to send in a request
+	maxBatchSize = 20
+	// retryMaxBatchSize is the max number of writes to send in a retry request
+	retryMaxBatchSize = 10
+	// maxRetryAttempts is the max number of times to retry a write
+	maxRetryAttempts = 10
+	// defaultStartingMaximumOpsPerSecond is the starting max number of requests to the service per second
+	defaultStartingMaximumOpsPerSecond = 500
+)
+
+// BulkWriterJob provides read-only access to the results of a BulkWriter write attempt.
+type BulkWriterJob struct {
+	err      chan error           // send error responses to this channel
+	result   chan *pb.WriteResult // send the write results to this channel
+	write    *pb.Write            // the writes to apply to the database
+	attempts int                  // number of times this write has been attempted
+	wr       *WriteResult         // (cached) result from the operation
+	e        error                // (cached) any errors that occurred
+	ctx      context.Context      // context for canceling/timing out results
+}
+
+// Results gets the results of the BulkWriter write attempt.
+// This method blocks if the results for this BulkWriterJob haven't been
+// received.
+func (j *BulkWriterJob) Results() (*WriteResult, error) {
+	if j.wr == nil && j.e == nil {
+		j.wr, j.e = j.processResults() // cache the results for additional calls
```

Comment:
In general our clients are thread safe, but calling this function is not because of the caching feature. Is there a need to cache the values? What is the purpopse of being able to call this method more than once?

Reply:
It would be nice to avoid the caching, but I suspect `Results` could get called a few times.

Putting `j.wr` and `j.e` behind a mutex would be reasonable.

Reply:
Yeah, I'm thinking about the possibility of `Results()` being called more than once. I'll put these fields behind a mutex